crates/kornia-linalg/src/linalg.rs

```diff
@@ -0,0 +1,509 @@
+    R.y_axis.z = -b * B.x_axis.z + a * B.y_axis.z;
+    R.z_axis.x = B.z_axis.x;
+    R.z_axis.y = B.z_axis.y;
+    R.z_axis.z = B.z_axis.z;
+
+    // Second Givens rotation (ch, 0, -sh, 0)
+    let g2 = qr_givens_quaternion(R.x_axis.x, R.z_axis.x);
+    a = -2.0 * g2.sh * g2.sh + 1.0;
+    b = 2.0 * g2.ch * g2.sh;
+
+    // Apply B = Q' * B
+    B.x_axis.x = a * R.x_axis.x + b * R.z_axis.x;
+    B.x_axis.y = a * R.x_axis.y + b * R.z_axis.y;
+    B.x_axis.z = a * R.x_axis.z + b * R.z_axis.z;
+    B.y_axis.x = R.y_axis.x;
+    B.y_axis.y = R.y_axis.y;
+    B.y_axis.z = R.y_axis.z;
+    B.z_axis.x = -b * R.x_axis.x + a * R.z_axis.x;
+    B.z_axis.y = -b * R.x_axis.y + a * R.z_axis.y;
+    B.z_axis.z = -b * R.x_axis.z + a * R.z_axis.z;
+
+    // Third Givens rotation (ch, sh, 0, 0)
+    let g3 = qr_givens_quaternion(B.y_axis.y, B.z_axis.y);
+    a = -2.0 * g3.sh * g3.sh + 1.0;
+    b = 2.0 * g3.ch * g3.sh;
+
+    // R is now set to desired value
+    R.x_axis.x = B.x_axis.x;
+    R.x_axis.y = B.x_axis.y;
+    R.x_axis.z = B.x_axis.z;
+    R.y_axis.x = a * B.y_axis.x + b * B.z_axis.x;
+    R.y_axis.y = a * B.y_axis.y + b * B.z_axis.y;
+    R.y_axis.z = a * B.y_axis.z + b * B.z_axis.z;
+    R.z_axis.x = -b * B.y_axis.x + a * B.z_axis.x;
+    R.z_axis.y = -b * B.y_axis.y + a * B.z_axis.y;
+    R.z_axis.z = -b * B.y_axis.z + a * B.z_axis.z;
+
+    // Construct the cumulative rotation Q = Q1 * Q2 * Q3
+    let sh12 = 2.0 * (g1.sh * g1.sh - 0.5);
+    let sh22 = 2.0 * (g2.sh * g2.sh - 0.5);
+    let sh32 = 2.0 * (g3.sh * g3.sh - 0.5);
+
+    Q.x_axis.x = sh12 * sh22;
+    Q.x_axis.y = 4.0 * g2.ch * g3.ch * sh12 * g2.sh * g3.sh + 2.0 * g1.ch * g1.sh * sh32;
+    Q.x_axis.z = 4.0 * g1.ch * g3.ch * g1.sh * g3.sh - 2.0 * g2.ch * sh12 * g2.sh * sh32;
+
+    Q.y_axis.x = -2.0 * g1.ch * g1.sh * sh22;
+    Q.y_axis.y = -8.0 * g1.ch * g2.ch * g3.ch * g1.sh * g2.sh * g3.sh + sh12 * sh32;
+    Q.y_axis.z =
+        -2.0 * g3.ch * g3.sh + 4.0 * g1.sh * (g3.ch * g1.sh * g3.sh + g1.ch * g2.ch * g2.sh * sh32);
+
+    Q.z_axis.x = 2.0 * g2.ch * g2.sh;
+    Q.z_axis.y = -2.0 * g3.ch * sh22 * g3.sh;
+    Q.z_axis.z = sh22 * sh32;
+
+    QR { Q, R }
+}
+
+/// Wrapping function used to contain all of the required sub calls
+pub fn svd(A: Mat3) -> SVDSet {
+    // Compute the eigenvectors of A^T * A, which is V in SVD (Singular Vectors)
+    let V = jacobi_eigenanalysis(Symmetric3x3::from_mat3x3(&(A.transpose().mul_mat3(&A))));
+
+    // Compute B = A * V
+    let mut B = A.mul_mat3(&V);
+
+    // Sort the singular values
+    sort_singular_values(&mut B, &mut V.clone());
+
+    // Perform QR decomposition on B to get Q and R
+    let qr = qr_decomposition(&mut B);
+
+    // Reset MXCSR register (if needed)
+
+    // Return the SVD result, which includes Q (as U), R (as S), and V
+    SVDSet {
+        U: qr.Q,
+        S: qr.R,
+        V,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use glam::{Mat3, Vec3};
+
+    use super::*;
+
+    #[test]
+    fn test_svd() {
```

Comment:
can you add a numerical test ? to check that we do not have any regression ? You could als try to reconstuct the input `A' = U @ (S @ VT)`